syntax = "proto3";

package grpcServer;

import "protos/product/product.proto";

option go_package = "github.com/lessbutter/alloff-api/api/grpcServer";

service ProductGroup {
    rpc GetProductGroup(GetProductGroupRequest) returns (GetProductGroupResponse);
    rpc CreateProductGroup(CreateProductGroupRequest) returns (CreateProductGroupResponse);
    rpc ListProductGroups(ListProductGroupsRequest) returns (ListProductGroupsResponse);
    rpc EditProductGroup(EditProductGroupRequest) returns (EditProductGroupResponse);
    rpc PushProducts(PushProductsRequest) returns (PushProductsResponse);
}

/*
    GRPC API Request & Response
*/

// Product Group Service
message GetProductGroupRequest {
    string product_group_id = 1;
}
message GetProductGroupResponse {
    ProductGroupMessage pg = 1;
}
message CreateProductGroupRequest {
    string title = 1;
    string short_title = 2;
    repeated string instruction = 3;
    string image_url = 4;
    string start_time = 5;
    string finish_time = 6;
}
message CreateProductGroupResponse {
    ProductGroupMessage pg = 1;
}
message EditProductGroupRequest {
    optional string title = 1;
    optional string short_title = 2;
    repeated string instruction = 3;
    optional string image_url = 4;
    optional string start_time = 5;
    optional string finish_time = 6;
    string product_group_id = 7;
}
message EditProductGroupResponse {
    ProductGroupMessage pg = 1;
}
message ListProductGroupsRequest {}
message ListProductGroupsResponse {
    repeated ProductGroupMessage pgs = 1;
}
message PushProductsRequest {
    string product_group_id = 1;
    repeated string product_id = 2;
    int32 priority = 3;
}
message PushProductsResponse {
    ProductGroupMessage pg = 1;
}

/*
    GRPC Structs
*/

message ProductInGroupMessage {
    ProductMessage product = 1;
    int32 priority = 2;
}

message ProductGroupMessage {
    string title = 1;
    string short_title = 2;
    repeated string instruction = 3;
    string image_url = 4;
    repeated ProductInGroupMessage products = 5;
    string start_time = 6;
    string finish_time = 7;
    string product_group_id = 8;
}
